src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -282,11 +282,36 @@ QueryPlanPtr MergeTreeDataSelectExecutor::readFromParts(
 
     auto query_context = context->hasQueryContext() ? context->getQueryContext() : context;
 
+    PartFilterCounters part_filter_counters;
+    auto index_stats = std::make_unique<ReadFromMergeTree::IndexStats>();
+
     if (query_context->getSettingsRef().allow_experimental_query_deduplication)
-        selectPartsToReadWithUUIDFilter(parts, part_values, minmax_idx_condition, minmax_columns_types, partition_pruner, max_block_numbers_to_read, query_context);
+        selectPartsToReadWithUUIDFilter(parts, part_values, minmax_idx_condition, minmax_columns_types, partition_pruner, max_block_numbers_to_read, query_context, part_filter_counters);
     else
-        selectPartsToRead(parts, part_values, minmax_idx_condition, minmax_columns_types, partition_pruner, max_block_numbers_to_read);
+        selectPartsToRead(parts, part_values, minmax_idx_condition, minmax_columns_types, partition_pruner, max_block_numbers_to_read, part_filter_counters);
+
+    index_stats->emplace_back(ReadFromMergeTree::IndexStat{
```

Comment:
Maybe we can avoid printing it if we have at least one index condition?

Reply:
I think it is useful to add.
We can see how indexes drop parts and granules.